add Jenkins with build/test functionality and GitHub postback

kibbles-n-bytes opened the pull request:
With this, by adding the 'jenkins' label to your PR, Jenkins will run tests (currently only Travis equivalency) on it.

duglin commented on contrib/hack/kubectl:
```diff
@@ -34,7 +34,7 @@ if [ -n "${PORT}" ]; then
 	SERVER_INFO="--server localhost:${PORT}"
 fi
 
-docker run --rm -ti --net host \
```
Not sure this is right. If kubectl ever needs input from the user (e.g. kubectl run --image ubuntu bash) then w/o `-ti` things won't work.

kibbles-n-bytes replied:
Jenkins does not support TTYs and breaks when you run docker commands that provision them. If we're using this script in the testing framework then it's necessary to remove the `-ti`. Most of what we'd do with this image seems to not require input, so it makes sense to me to deal with those specific use cases if they start to become prevalent.

If you feel it's important ASAP, we could add an environment variable flag to this script that specifies whether the command we're running needs a TTY (so you'd call `WITH_TTY=1 kubectl run $ARGS`).

kibbles-n-bytes commented:
@duglin I went ahead and added the environment variable flag for `kubectl`. I also changed the filename to `kubectl.sh`; I know the intent was to make calling kubectl via docker notationally equivalent to calling it without, but it also obscures the filetype, causing me problems when trying to debug the TTY issue as I was treating it as a black-box binary. Let me know your thoughts.

arschles commented:
@kibbles-n-bytes does this fix either #487 or #456?

kibbles-n-bytes commented:
@arschles No, this brings Jenkins up to Travis equivalency but it does not currently run the e2e tests, due to the changes in the controller and walkthrough I took that out temporarily and am working on adding that back in now.

duglin commented on contrib/hack/kubectl.sh:
```diff
@@ -17,6 +17,12 @@ set -e
 
 ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/../.." && pwd)"
 
+if [[ "${WITH_TTY:-}" -eq 1 ]]; then
```
If we're going to go this path then the default needs to be to have `-ti` enabled and let's make jenkins turn it off.
I want people to be able to run kubectl to test stuff out (manually) w/o having to set additional flag.s  The point of this script is to be just like the normal/stand-alone kubectl as much as possible.

duglin replied:
Along those line, I'm NOT LGTM on renaming this kubectl.sh - I want people to be able to add contrib/hack to their path and then run `kubectl` from whatever scripts they might have.

duglin replied:
Net- Jenkins is the odd ball, not a normal user  :-)

kibbles-n-bytes commented:
@duglin Continuing the discussion out here, since I pushed the change to un-default TTY-less, so GitHub collapsed the discussion.

Having the TTY-less version be the alternative is fine by me, I was just hesitant to pollute `setup-kubectl.sh` and `test-apiserver.sh` with a bunch of "no tty" statements. I agree Jenkins is the odd-ball.

For `kubectl` vs. `kubectl.sh`, I feel like that's solved just as easily by making a simlink called `kubectl` and pointing it to `kubectl.sh`. And that way, it's clearer to both the user and anyone looking at our code what this file is actually doing. This isn't a Jenkins issue; this is just something that was confusing to me in general, as looking at the filename and extension, you expect it to be an executable, but it's actually a shell script.

kibbles-n-bytes commented:
@duglin That said, I'll go ahead and punt this for now, rename it, unblock this PR so we can get Jenkins in in some manner, and then open up a separate CL to see people's thoughts.

duglin commented on contrib/hack/setup-kubectl.sh:
```diff
@@ -46,8 +46,8 @@ D_HOST=${D_HOST#*//}   # remove leading proto://
-kubectl config set-cluster service-catalog-cluster --server=http://${D_HOST}:${PORT}
-kubectl config set-context service-catalog-ctx --cluster=service-catalog-cluster --user=service-catalog-creds
-kubectl config use-context service-catalog-ctx
+NO_TTY=1 kubectl config set-credentials service-catalog-creds --username=admin --password=admin
```
Minor thing, but here and below in the test, you could just do `export NO_TTY=1` on a line by itself before all these calls to kubectl and it should pick it up, I think.

duglin commented:
LGTM

would prefer to not have NO_TTY=1 on each line (just a bit verbose) though.

arschles commented:
@kibbles-n-bytes does this close https://github.com/kubernetes-incubator/service-catalog/issues/487 and/or https://github.com/kubernetes-incubator/service-catalog/issues/456? If so, can you add `Closes #12345` into the OP?

arschles commented:
nevermind, I saw the answer above. Merging.
